src/ui/auth/rackspace/rackspace.go

```diff
@@ -0,0 +1,174 @@
+	} `json:"status"`
+}
+
+var (
+	rackspaceMK8SAuthURLTokenEndpoint string
+)
+
+// Authenticate checks user's credential against the Rackspace Managed Kubernetes Auth (kubernetes-auth)
+// if the check is successful a dummy record will be inserted into DB, such that this user can
+// be associated to other entities in the system.
+func (l *Auth) Authenticate(m models.AuthModel) (*models.User, error) {
+
+	// kubernetes-auth only uses the token (m.Password) for auth. The username (m.Principal) isn't used at all.
+	// In fact, a user could put anything at all into the username field. It must be ignored.
+	// However, we log the username to help track the request because we can't put the token in the logs.
+	log.Debugf("ProvidedUsername=%s Authentication attempt", m.Principal)
+
+	// build auth request
+	authRequestBody := fmt.Sprintf("{\"spec\": {\"token\": \"%s\"}}", m.Password)
+	req, err := http.NewRequest(http.MethodPost, rackspaceMK8SAuthURLTokenEndpoint, strings.NewReader(authRequestBody))
+	if err != nil {
+		log.Errorf("ProvidedUsername=%s Error=%v", m.Principal, err)
+		return nil, err
+	}
+
+	log.Debugf("ProvidedUsername=%s Sending auth request to %s", m.Principal, rackspaceMK8SAuthURLTokenEndpoint)
+
+	// send auth request
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Errorf("ProvidedUsername=%s Error=%v", m.Principal, err)
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	// read auth response body
+	authRespBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Errorf("ProvidedUsername=%s Error=%v", m.Principal, err)
+		return nil, err
+	}
+
+	// check for auth server error
+	if resp.StatusCode >= http.StatusInternalServerError {
+		log.Errorf("ProvidedUsername=%s Error=%s", m.Principal, authRespBody)
+		return nil, fmt.Errorf("%d %s", resp.StatusCode, authRespBody)
+	}
+
+	// read auth response body as json
+	authResp := AuthResponse{}
+	err = json.Unmarshal([]byte(authRespBody), &authResp)
+	if err != nil {
+		log.Errorf("ProvidedUsername=%s Error=%v", m.Principal, err)
+		return nil, err
+	}
+
+	log.Debugf("ProvidedUsername=%s Authenticated=%t", m.Principal, authResp.Status.Authenticated)
+
+	// check for any status other than OK
+	if resp.StatusCode != http.StatusOK {
+		log.Errorf("ProvidedUsername=%s Error=%s", m.Principal, authResp.Status.Error)
+		return nil, fmt.Errorf("%d %s", resp.StatusCode, authResp.Status.Error)
+	}
+
+	log.Debugf("ProvidedUsername=%s UID=%s BackendUsername=%s", m.Principal, authResp.Status.User.UID, authResp.Status.User.Username)
+
+	// check if the user already exists in the Harbor DB. if the user doesn't exist, create it.
+	user := models.User{}
+	// set the Harbor Realname to the kubernetes-auth backend's UID because the UID is a static ID
+	// whereas the kubernetes-auth backend's Username can change (so put it in the Harbor Username field for convenience)
+	user.Realname = authResp.Status.User.UID
+
+	exist, err := dao.UserExists(user, "realname")
```

Comment:
s/exist/exists/

Reply:
to improve performance (1 mysql call instead of 2), you can do:

```
user, err := dao. GetUser(user, "realname")
...
if user == nil {
   // doesn't exist
} 
```